### PR TITLE
Messages User Guide

### PR DESCRIPTION
**Pull request by sougou**

### CODE REVIEW

File: doc/Messaging.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+The above indexes are recommended for optimum performance. However, some variation can be allowed to achieve different performance trade-offs.
+
+The comment section specifies additional configuration parameters. The fields are as follows:
+
+* `vitess_message`: Indicates that this is a message table.
+* `vt_ack_wait=30`: Wait for 30s for the first message ack. If one is not received, resend.
+* `vt_purge_after=86400`: Purge acked messages that are older than 86400 seconds (1 day).
+* `vt_batch_size=10`: Send up to 10 messages per batch.
+* `vt_cache_seze=10000`: Store up to 10000 messages in the cache.
```
Comment (derekperkins) [inline doc/Messaging.md]: should be `vt_cache_size`

File: doc/Messaging.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+# Enqueuing messages
+
+The application can enqueue messages using an insert statement:
+
+```
+insert into my_message(id, message) values(1, 'hello world')
+```
+
+These inserts can be part of a regular transaction. Multiple messages can be inserted to different tables. Avoid accumulating too many big messages within a transaction as it consumes memory on the VTTablet side. At the time of commit, memoery permitting, all messages are instantly enqueued to be sent.
```
Comment (derekperkins) [inline doc/Messaging.md]: memoery to memory

File: doc/Messaging.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+```
+create table my_message(
+  time_scheduled bigint,
+  id bigint,
+  time_next bigint,
+  epoch bigint,
+  time_created bigint,
+  time_acked bigint,
+  message varchar(128),
```
Comment (derekperkins) [inline doc/Messaging.md]: Why `varchar` instead of `varbinary`? Is the size configurable?

File: doc/Messaging.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+Messages are not a good fit for the following use cases:
+
+* Broadcasting of events to multiple subscribers.
+* Ordered delivery.
+* Real-time delivery.
+
+# Creating a message table
+
+The current implementation requires a fixed schema. This will be made more flexible in the future. There will also be a custom DDL syntax. For now, a message table must be created like this:
```
Comment (derekperkins) [inline doc/Messaging.md]: 1 table ~ 1 queue, so inserts / acks will be limited by your single node performance, correct?

File: doc/Messaging.md (hunk 1 above)
Comment (derekperkins) [inline doc/Messaging.md]: What are `vt_batch_size` and `vt_cache_size`?

File: doc/Messaging.md (hunk 2 above)
Comment (derekperkins) [inline doc/Messaging.md]: What does that VTTablet memory limitation look like? It's pretty common for us to add > 1M tasks in a single transaction, with the payload being a ~100 character url.

File: doc/Messaging.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+The message feature is currently in alpha, and can be improved. Here is the list of possible limitations/improvements:
+
+* Flexible columns: Allow any number of application defined columns to be in the message table.
+* No ACL check for receivers: To be added.
+* Proactive scheduling: Upcoming messages can be proactively scheduled for timely delivery instead of waiting for the next polling cycle.
+* Monitoring support: To be added.
+* Dropped tables: The message engine does not currently detect dropped tables.
+* Changed properties: Although the engine detects new message tables, it does not refresh properties of an existing table.
+* A `SELECT` style syntax for subscribing to messages.
+* No rate limiting.
```
Comment (derekperkins) [inline doc/Messaging.md]: Rate limiting would be my #1 request

File: doc/Messaging.md [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+update my_message set time_acked = :time_acked, time_next = null where id in ::ids and time_acked is null
+```
+
+You can also view messages using regular `select` queries.
+
+# Known limitations
+
+The message feature is currently in alpha, and can be improved. Here is the list of possible limitations/improvements:
+
+* Flexible columns: Allow any number of application defined columns to be in the message table.
```
Comment (derekperkins) [inline doc/Messaging.md]: Flexible columns is a close second on the feature list

File: doc/Messaging.md [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+* `Name`: Name of the message table.
+* `Keyspace`: Keypsace where the message table is present. This field can be empty if the table name is unique across all keyspaces.
+* `Ids`: The list of ids that need to be acked.
+
+Once a message is successfully acked, it will never be resent.
+
+## Exponential backoff
+
+A message that was successfully sent will wait for the specified ack wait time. If no ack is received by then, it will be resent. The next attempt will be 2x the previous wait, and this delay is doubled for every attempt.
```
Comment (derekperkins) [inline doc/Messaging.md]: Would it be possible to set a queue level min/max?

File: doc/Messaging.md [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+# Known limitations
+
+The message feature is currently in alpha, and can be improved. Here is the list of possible limitations/improvements:
+
+* Flexible columns: Allow any number of application defined columns to be in the message table.
+* No ACL check for receivers: To be added.
+* Proactive scheduling: Upcoming messages can be proactively scheduled for timely delivery instead of waiting for the next polling cycle.
+* Monitoring support: To be added.
+* Dropped tables: The message engine does not currently detect dropped tables.
+* Changed properties: Although the engine detects new message tables, it does not refresh properties of an existing table.
```
Comment (derekperkins) [inline doc/Messaging.md]: Does properties mean the columns or the rows themselves? Is it possible to `UPDATE` the table? For example, sometimes we've had situations where exponential backoff has pushed the next execution time back hours or days, but it is especially helpful during debugging to have a message run now, and then update all those messages to run now after the bug has been fixed.

File: doc/Messaging.md (hunk 6 above)
Comment (sougou) [inline doc/Messaging.md]: This won't be too hard to add on (after I take care of a few unrelated high-priority items).

**Comment by sougou** [comment]
About your comment related to making your message rows co-exist with a parent row of another table, you can still get that using a custom sharding scheme. However, it could lead to hotspots if some root rows generate more messages than others.

Let's say the parent table has `user_id` as the primary key, the message table can have an `id` column that contains the `user_id` in its bits, and you can have it use the same sharding, but only using the user_id parts as input.

**Comment by sougou** [comment]
Ping. Need lgtm.

File: doc/Messaging.md [hunk 9]
```diff
@@ -0,0 +1,148 @@
+# Vitess Messaging
+
+# Overview
+
+Vitess messaging gives the application an easy way to schedule and manage work that needs to be performed asynchronously. Under the covers, messages are stored in a traditional MySQL table and therefore enjoy the following properties:
```
Comment (alainjobart) [inline doc/Messaging.md]: Can you format this file with < 80 characters per line? My emacs does that for me...

File: doc/Messaging.md [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+* **Transactional**: Messages can be created or acked as part of an existing transaction. The action will complete only if the commit succeeds.
```
Comment (alainjobart) [inline doc/Messaging.md]: This one is not the main attraction, I would move it down, keep Scalable as the first one in the list.

**Comment by alainjobart** [comment]
LGTM

**Pull request merged**